src/main/java/graphql/servlet/context/ContextSetting.java

```diff
@@ -0,0 +1,95 @@
+package graphql.servlet.context;
+
+import graphql.ExecutionInput;
+import graphql.execution.ExecutionId;
+import graphql.execution.instrumentation.ChainedInstrumentation;
+import graphql.execution.instrumentation.Instrumentation;
+import graphql.execution.instrumentation.dataloader.DataLoaderDispatcherInstrumentationOptions;
+import graphql.schema.GraphQLSchema;
+import graphql.servlet.input.GraphQLBatchedInvocationInput;
+import graphql.servlet.input.PerQueryBatchedInvocationInput;
+import graphql.servlet.input.PerRequestBatchedInvocationInput;
+import graphql.servlet.instrumentation.ConfigurableDispatchInstrumentation;
+import graphql.servlet.instrumentation.FieldLevelTrackingApproach;
+import graphql.servlet.instrumentation.RequestLevelTrackingApproach;
+import graphql.servlet.core.internal.GraphQLRequest;
+import org.dataloader.DataLoaderRegistry;
+
+import java.util.Arrays;
+import java.util.List;
+
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * An enum representing possible context settings. These are modeled after Apollo's link settings.
+ */
+public enum ContextSetting {
+
+    /**
+     * A context object, and therefor dataloader registry and subject, should be shared between all GraphQL executions in a http request.
+     */
+    PER_REQUEST_WITH_INSTRUMENTATION,
+    PER_REQUEST_WITHOUT_INSTRUMENTATION,
+    /**
+     * Each GraphQL execution should always have its own context.
+     */
+    PER_QUERY_WITH_INSTRUMENTATION,
+    PER_QUERY_WITHOUT_INSTRUMENTATION;
+
+    /**
+     * Creates a set of inputs with the correct context based on the setting.
+     * @param requests the GraphQL requests to execute.
+     * @param schema the GraphQL schema to execute the requests against.
+     * @param contextSupplier method that returns the context to use for each execution or for the request as a whole.
+     * @param root the root object to use for each execution.
+     * @return a configured batch input.
+     */
+    public GraphQLBatchedInvocationInput getBatch(List<GraphQLRequest> requests, GraphQLSchema schema, Supplier<GraphQLContext> contextSupplier, Object root) {
+        switch (this) {
+            case PER_QUERY_WITH_INSTRUMENTATION:
+                //Intentional fallthrough
+            case PER_QUERY_WITHOUT_INSTRUMENTATION:
+                return new PerQueryBatchedInvocationInput(requests, schema, contextSupplier, root);
+            case PER_REQUEST_WITHOUT_INSTRUMENTATION:
+                //Intentional fallthrough
+            case PER_REQUEST_WITH_INSTRUMENTATION:
+                return new PerRequestBatchedInvocationInput(requests, schema, contextSupplier.get(), root);
```

Comment:
I think I get the difference between PerQueryBatchedInvocationInput and PerRequestBatchedInvocationInput, but I think it would difficult for contextSupplier to provide different context for PerQueryBatchedInvocationInput, since the contextSupplier does not have visibility into which query the context is required.

Reply:
It would be up to the person customizing the ContextBuilder to keep in mind which setting they are going to be using. Do you think they need some part of the input object that they can't get from the http objects?

Reply:
Also the intent for the context isn't that they are based on any property of the query, so I'm not sure I see the problem